logpyle/__init__.py

```diff
@@ -650,6 +645,13 @@ def __init__(self, filename: Optional[str] = None, mode: str = "r",
 
         # }}}
 
+        # {{{ atexit handling
+
+        if self.mode[0] == "w":
+            atexit.register(self.save)
```

Comment:
Doesn't this *also* keep the instance alive? (Maybe use a weak reference?)

Reply:
[c9777fc](https://github.com/illinois-ceesd/logpyle/pull/146/commits/c9777fc68ea282b838455b7b75103e879d56035e) changes this to a weak reference. I couldn't find a way to both (a) delete the instance immediately when `del` is called and (b) ensure that `self.save` is called on the various "unclean-exit" scenarios.

Reply:
> delete the instance immediately when del is called

Wait: not sure I understand this. Doesn't a call to `__del__` immediately precede instance deletion?

Reply:
What I meant was that calling `del logmgr` does not immediately call `logmgr.__del__`, due to the `weakref.finalize` holding a reference to `logmgr.save` (afaict; not creating the `weakref.finalize` causes `del logmgr` to immediately run `logmgr.__del__`). I tested `weakref.{proxy,ref}` and they seem to have the same behavior. `logmgr.__del__` does get called and runs as expected when Python exits.

See also the comment at https://github.com/illinois-ceesd/logpyle/pull/146/files#diff-869af26f06e5c0fd42958eee7b7000279506664c8887ac7f0fc958bf9c20867bR843-R847

Reply:
I still think this should be possible, but it's perhaps not the highest priority. File a bug? If this improves behavior for mirgecom, then I'd say go ahead.

Reply:
Yeah, I agree. I left this as a FIXME in the code. This particular scenario (`del` without `close`) shouldn't happen in normal operation.